x/keyshare/keeper/msg_server_submit_encrypted_keyshare.go

```diff
@@ -0,0 +1,143 @@
+package keeper
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"cosmossdk.io/math"
+	commontypes "github.com/Fairblock/fairyring/x/common/types"
+	"github.com/Fairblock/fairyring/x/keyshare/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+)
+
+func (k msgServer) SubmitEncryptedKeyshare(goCtx context.Context, msg *types.MsgSubmitEncryptedKeyshare) (*types.MsgSubmitEncryptedKeyshareResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// check if validator is registered
+	_, found := k.GetValidatorSet(ctx, msg.Creator)
+
+	if !found {
+		authorizedAddrInfo, found := k.GetAuthorizedAddress(ctx, msg.Creator)
+		if !found || !authorizedAddrInfo.IsAuthorized {
+			return nil, types.ErrAddrIsNotValidatorOrAuthorized.Wrap(msg.Creator)
+		}
+
+		_, found = k.GetValidatorSet(ctx, authorizedAddrInfo.AuthorizedBy)
+		if !found {
+			return nil, types.ErrAuthorizerIsNotValidator.Wrap(authorizedAddrInfo.AuthorizedBy)
+		}
+
+		// If the sender is in the validator set & authorized another address to submit key share
+	} else if count := k.GetAuthorizedCount(ctx, msg.Creator); count != 0 {
+		return nil, types.ErrAuthorizedAnotherAddress
+	}
+
+	keyShareReq, found := k.GetPrivateKeyShareRequest(ctx, msg.Identity)
+	if !found {
+		return nil, types.ErrKeyShareRequestNotFound.Wrapf(", got id value: %s", msg.Identity)
+	}
+
+	valEncKeyshare := types.ValidatorEncryptedKeyShare{
+		Validator:           msg.Creator,
+		Requester:           msg.Requester,
+		KeyShare:            msg.EncryptedKeyshare,
+		KeyShareIndex:       msg.KeyShareIndex,
+		ReceivedTimestamp:   msg.ReceivedTimestamp,
+		ReceivedBlockHeight: msg.ReceivedBlockHeight,
+		Identity:            msg.Identity,
+	}
+
+	// Save the new general key share to state
+	k.SetPrivateKeyShare(ctx, valEncKeyshare)
+	k.SetLastSubmittedHeight(ctx, msg.Creator, strconv.FormatInt(ctx.BlockHeight(), 10))
+
+	validatorList := k.GetAllValidatorSet(ctx)
+
+	// Get all the general key shares for the provided id value & id type
+	var stateEncryptedKeyShares []types.ValidatorEncryptedKeyShare
+
+	for _, eachValidator := range validatorList {
+		eachEncKeyShare, found := k.GetPrivateKeyShare(ctx, eachValidator.Validator, msg.Identity, msg.Requester)
+		if !found {
+			continue
+		}
+		stateEncryptedKeyShares = append(stateEncryptedKeyShares, eachEncKeyShare)
+	}
+
+	// Get the active public key for aggregating
+	activePubKey, found := k.GetActivePubKey(ctx)
+
+	if !found {
+		return nil, types.ErrPubKeyNotFound
+	}
+
+	expectedThreshold := math.LegacyNewDecFromInt(
+		math.NewInt(types.KeyAggregationThresholdNumerator)).Quo(
+		math.LegacyNewDecFromInt(math.NewInt(types.KeyAggregationThresholdDenominator))).MulInt64(
+		int64(activePubKey.NumberOfValidators)).Ceil().TruncateInt64()
+
+	// Emit KeyShare Submitted Event
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(types.SendGeneralKeyshareEventType,
```

Comment:
You didn't update the event type & attributes

Reply:
updated event